File: python/141. Linked list Cycle/linked_list_cycle.py
from typing import Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def has_cycle(head: Optional[ListNode]) -> bool:
    slow = head
    fast = head

    while(fast and fast.next):
        slow = slow.next
        fast = fast.next.next

        if (fast == slow): return True

    return False


if __name__ == '__main__':
    head = ListNode(3)
    sec = ListNode(2)
    thir = ListNode(0)
    four = ListNode(-4)

    head.next = sec
    sec.next = thir
    thir.next = four

    # Without a cycle
    print(has_cycle(head))

    # With a cycle
    four.next = sec

    print(has_cycle(head))

File: c++/141. Linked list Cycle/linked_list_cycle.cpp
#include<iostream>
using namespace std;


struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};


bool hasCycle(ListNode* head) {
    ListNode* slow = head;
    ListNode* fast = head;

    while(fast != NULL) {
        fast = fast->next;
        if(fast != NULL) {
            fast = fast->next;
            slow = slow->next;
        }

        if(fast == slow) {
            return true;
        }
    }
    return false;
}

int main() {
    ListNode* head = new ListNode(3);
    ListNode* sec = new ListNode(2);
    ListNode* thir = new ListNode(0);
    ListNode* four = new ListNode(-4);

    head->next = sec;
    sec->next = thir;
    thir->next = four;

    // Without a cycle
    cout<<hasCycle(head)<<endl;

    // With a cycle
    four->next = sec;
    cout<<hasCycle(head)<<endl;

    return 0;
}